userspace/libsinsp/event.h

```diff
@@ -91,25 +92,81 @@ class SINSP_PUBLIC sinsp_evttables
 	const struct ppm_event_info* m_event_info; ///< List of events supported by the capture and analysis subsystems. Each entry fully documents an event and its parameters.
 };
 
+template<class T> inline T get_event_param_as(const class sinsp_evt_param& param);
+
 /*!
   \brief Event parameter wrapper.
-  This class describes a raw event coming from the driver.
+  This class describes an event parameter coming from the driver.
 */
 class SINSP_PUBLIC sinsp_evt_param
 {
 public:
-	char* m_val;	///< Pointer to the event parameter data.
+	const sinsp_evt *m_evt; ///< Pointer to the event that contains this param
+	uint32_t m_idx; ///< Index of the parameter within the event
+
+	const char* m_val;	///< Pointer to the event parameter data.
 	uint32_t m_len; ///< Length of the parameter pointed by m_val.
-private:
-	inline void init(char* valptr, uint32_t len)
+
+	sinsp_evt_param(const sinsp_evt *evt, uint32_t idx, const char *val, uint32_t len):
+		m_evt(evt), m_idx(idx), m_val(val), m_len(len) {}
+
+	/*!
+	  \brief Interpret the parameter as a specific type, like:
+	  	- Fixed size values (uint32_t, int8_t ..., e.g. param->as<uint32_t>())
+		- String-like types with std::string_view (e.g. param->as<std::string_view>())
+	*/
+	template<class T>
+	inline T as() const
 	{
-		m_val = valptr;
-		m_len = len;
+		return get_event_param_as<T>(*this);
 	}
 
-	friend class sinsp_evt;
+	const struct ppm_param_info* get_info() const;
+	std::string invalid_len_error(size_t requested_len) const;
 };
 
+/*!
+  \brief Get the value of a parameter, interpreted with the type specified in the template argument.
+  \param param The parameter.
+*/
+template<class T>
+inline T get_event_param_as(const sinsp_evt_param& param)
+{
+	T ret;
+
+	if (param.m_len != sizeof(T))
+	{
+		// By moving this error string building operation to a separate function
+		// the compiler is more likely to inline this entire function.
+		// This is important since get_param<> is called in the hot path.
+		throw sinsp_exception(param.invalid_len_error(sizeof(T)));
+	}
+
+	memcpy(&ret, param.m_val, sizeof(T));
+
+	return ret;
+}
+
+template<>
+inline std::string_view get_event_param_as<std::string_view>(const sinsp_evt_param& param)
+{
+	if (param.m_len == 0)
+	{
+		return {};
+	}
+
+	size_t string_len = strnlen(param.m_val, param.m_len);
+	if (param.m_len == string_len)
```

Comment:
It is indeed a NUL terminator check. Sometimes the drivers will output a NUL terminated string and I feel we need a type for that, and that's what I'm using `std::string_view` for (i.e., this will interpret one parameter that contains a NUL-terminated string).

The resulting string view will not contain the null terminator within its bounds but if you call `.data()` it will return a NUL-terminated `const char*`. We wouldn't have this additional requirement if we were using `std::string` around libsinsp, but many functions right now do ask for a `const char*`.

This works, but does not feel ideal, I'm open to suggestions.

Reply:
So we're technically doing an out of bounds read when treating the string_view as a NUL-terminated string (but it's fine because we know there's a NUL just outside), right?

I'm fine with this (this-is-fine.jpg) but:
* please add a comment about it
* I'd explicitly check `param.m_val[param.m_len] == 0` instead (feels more direct) but that does change the behavior for NUL-separated string arrays (which we use in a few places) -- we'd start accepting them as string_views. I assume we don't want them as string views but I could also use a comment to that effect :)

Reply:
> doing an out of bounds read
Indeed we read within the event but outside the `string_view` which we guarantee to be NUL-terminated just outside. If we were using a NUL-terminated `string_view` we would have troubles converting to `std::string` because the terminator would be added to the string itself. Unless you can think of a way of having both right at the same time. The other alternative is to just copy the string.

Anyways I agree with you that the right way of fixing it is to avoid `const char*` as much as possible and use strings instead.

I have clarified the comment and changed the error condition, I think we simply expect the string to occupy the entire value except for the null terminator, all other cases should be handled differently.

Reply:
no, no, I do not advocate for NUL-terminated string views, using the NUL just outside it is a great hack in the circumstances we're in ;)

:+1: